src/Data/String/CodePoints.js

```diff
@@ -0,0 +1,119 @@
+"use strict";
+/* global Symbol */
+
+var hasArrayFrom = typeof Array.from === "function";
+var hasStringIterator =
+  typeof Symbol !== "undefined" &&
+  Symbol != null &&
+  typeof Symbol.iterator !== "undefined" &&
+  typeof String.prototype[Symbol.iterator] === "function";
+var hasFromCodePoint = typeof String.prototype.fromCodePoint === "function";
+var hasCodePointAt = typeof String.prototype.codePointAt === "function";
+
+exports._unsafeCodePointAt0 = function (fallback) {
+  return hasCodePointAt
+    ? function (str) { return str.codePointAt(0); }
+    : fallback;
+};
+
+exports._codePointAt = function (fallback) {
+  return function (Just) {
+    return function (Nothing) {
+      return function (unsafeCodePointAt0) {
+        return function (index) {
+          return function (str) {
+            var length = str.length;
+            if (index < 0 || index >= length) return Nothing;
+            if (hasArrayFrom) {
```

Comment:
You're right, the order could be reversed here. Maybe we should do some testing and use a heuristic? The string iterator is inherently slow and hard to optimise, so I'm sure there exists a ratio between `index` and `length` such that indexing into the result of `Array.from` ends up being faster. I just don't know what that is. It will also probably vary a good bit as engines work hard to optimise iterator usage patterns.